### ISSUE TITLE
[Tree] persistAs*SiblingOf without TreeRoot and no parent saves as last sibling anyway

### ISSUE THREAD

**Issue by Ymox** [Atlantic18/DoctrineExtensions#1481]
Hello everyone !

In my Symfony project, I have an entity that uses the NestedTree strategy, without `@Gedmo\TreeRoot`, to have a list of tasks (and incidentally subtasks) ordered. As those tasks are linked to a Project entity, this lets me prioritize not only tasks, but projects as well, even indirectly.

At a certain point, I display only tasks for a particular project, and I create some as well. For subtasks, I don't have any problem, but for tasks at level 0, I do.  
The main goal would be to set the newly created task at top of the list _for the project_. Speaking with a family, this means to me saving the new task as the _previous sibling of_ the first one for this project.

So I made my code, and soon discovered that I might have something wrong, as everytime I create a new task, it is placed at the very end of the complete list. After a serial of various tests – including persisting "normally", flushing, persisting as previous sibling, flushing again), I've been digging through the code, and I saw things that puzzled me.

In Gedmo\Tree\Strategy\ORM\Nested::updateNode, I see a complex treatment if there is a parent. In my case, of course, there's none, and I fell in [the second case](/Atlantic18/DoctrineExtensions/blob/master/lib/Gedmo/Tree/Strategy/ORM/Nested.php#L388).

It looks like this case assumes there's no way to know where to save the node, and therefore puts it at the end of the set. But we have the `sibling` property, [set in NestedTreeRepository](/Atlantic18/DoctrineExtensions/blob/master/lib/Gedmo/Tree/Entity/Repository/NestedTreeRepository.php#L113), so we don't need to fetch the "end of the tree". In this case, we simply can save the sibling's TreeLeft value, move the sibling down, and set the right and left weights to the new node according to the saved TreeLeft number. If the sibling has a parent, we can as well have the information from the sibling.
Similar for `persistAsNextSiblingOf`, where we have to save the TreeRight instead, and move down only the "next" ones,

I therefore wondered if I wasn't torturing myself and tried to add TreeRoot, as my case would be similar to multiple roots. But at this moment I get the error "Cannot persist sibling for a root node, tree operation is not possible".

Now's what's puzzling me : which difficulty did I miss that makes blocking the implementation of persisting a new node as a sibling of a given root node, with or without TreeRoot?

_Edit_ : I read #208, this feels like a duplicate, sorry. I'd continue over there in case it is better.
Meanwhile, I'll try to understand the parent problem.
The parent property is only here as help, so as the root property. Why would one of these be needed when the Nested Set model uses only the right and left values?

**Comment by l3pp4rd** [comment]
Hi, indeed, this is an old issue. I agree that if possible it should update the root node left right values to match the ordering. The problem is that I do not work with PHP or doctrine for some time and there is no interest or need to look into it. If that bothers you, maybe it would not be too hard to contribute and help others struggling with the same issue?

**Comment by Ymox** [comment]
Thanks for your reply.

I'll have a try at changing/adapting for those cases. Before starting, I just wanted to know if there was some peculiar data that makes them difficult to handle.

**Comment by l3pp4rd** [comment]
yes, currently I do not remember whether this issue was missed or was a reason to not to implement it.. if you have time to improve it, would be great ;)

**Comment by stof** [comment]
Well, you cannot have a sibling of a root node in a tree: the root node does not have any sibling, by definition of a tree.

**Comment by Ymox** [comment]
To sum up : if I want to order elements in a nested set, I need to

- either have multiple roots, but in this case I can't order them (new roots are "weighted" 1-2 apparently), which I need
- or create a root and duplicate its ID in every descendant at any level, all that to have weights for each tasks as I currently have

I remember having tested the last one, but I got the message that I couldn't persist as sibling of a root node, I'll try again.

**Comment by stof** [comment]
@Ymox you can order elements inside a tree. If they don't belong to the same tree, they cannot be ordered together.
If your items must be siblings to each other in the tree, it means they cannot be the root node. They need to have a parent.

The ``TreeRoot`` annotation is here to be able to store multiple **independant** trees in the same table (for instance having a separate tree per project). If you don't use this annotation, it means you have a single tree in the table and all items belong to it.
In any case, a tree can only have 1 root node, meaning you can only have 1 item without parent in each tree (and you must have one, as long as the tree is not empty)

**Comment by l3pp4rd** [comment]
@stof if tree has tree root then yes sure, but I think @Ymox uses tree without TreeRoot annotation and in that case **lft, rgt** values make sense to make ordering..
Also it would probably totally make sense to have TreeRoot as relation, which is another issue..

**Comment by stof** [comment]
@l3pp4rd but in this case, he can only have 1 root node (otherwise, it is not a tree anymore), so the tasks he want to order relatively to each other should have a parent

**Comment by l3pp4rd** [comment]
Yes that makes sense @stof probably that was the reason why it is not working for you @Ymox

**Comment by Ymox** [comment]
I agree that a tree needs a root.

Now if an ordered hierarchy without a root is not a tree, what is it then? And why consider it differently, just speaking about data layout and manipulation?

**Comment by stof** [comment]
@Ymox the notion of sliblings is easy: 2 nodes are siblings if they have the same parent. If you define a node without a parent, it cannot have siblings. The issue you have is related to the way you architectured your storage of tasks, as you tried to stored them in a nested set tree without architecturing them as a proper tree.

And btw, storing multiple root nodes (i.e. multiple nodes without parent) without using ``TreeRoot`` is not supported by the NestedSet strategy (due to the way storing the tree relationships is done). the Closure strategy does not need ``TreeRoot`` to store multiple trees in the same table due to the way data are represented (but children are not ordered when using the Closure strategy).

**Comment by l3pp4rd** [comment]
@Ymox have you tried **persistAsFirstChild** and **persistAsLastChild** ?

**Comment by Ymox** [comment]
All right, then. I added my root, and now quite everything is working. I'll just have to handle the fact that this root has not to show up in my case but used as parent by default.

Thanks to both of you for your time and work  :)
